### PR TITLE
fix #885: add default compilation unit when create a new CtClass

### LINKED ISSUE #885: IllegalStateException in JavaOutputProcessor#createJavaFile with the default source positioning
Dear all,

The default source positioning of a CtElement is created with a null compilation unit. If ones tries to create the Java file associated with such a element with JavaOutputProcessor#createJavaFile an IllegalStateException is raised.

The following test illustrates the issue:

https://github.com/seintur/spoon/blob/nullCompilationUnit_bug/src/test/java/spoon/test/sourcePosition/SourcePositionTest.java#L53

I'm not sure how to proceed.

In any case, IMHO, one should be able to generate the Java file for a CtClass created by the factory without having to bother about this.

Lionel.

### PR DESCRIPTION
**Pull request by danglotb**

### CODE REVIEW

File: src/test/java/spoon/test/prettyprinter/DefaultPrettyPrinterTest.java [hunk 1]
```diff
@@ -234,4 +237,22 @@ public void useFullyQualifiedNamesInCtElementImpl_toString() throws Exception {
...
+		l.getEnvironment().setNoClasspath(true);
+		l.buildModel();
+		Factory factory = l.getFactory();
+		CtClass<Object> ctClass = factory.Class().create("foo.Bar");
+		PrettyPrinter pp = new DefaultJavaPrettyPrinter(factory.getEnvironment());
+		JavaOutputProcessor jop =
+				new JavaOutputProcessor(File.createTempFile("foo","").getParentFile(),pp);
+		jop.setFactory(factory);
+		jop.createJavaFile(ctClass);
+		assertTrue(new File("/tmp/foo/Bar.java").exists());
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/prettyprinter/DefaultPrettyPrinterTest.java]: `File#createTempFile` will create the `Bar` class in the `/tmp` directory in all systems (Linux, MacOS and Windows)? I have doubts. (:

File: src/test/java/spoon/test/prettyprinter/DefaultPrettyPrinterTest.java [hunk 2]
```diff
@@ -234,4 +237,22 @@ public void useFullyQualifiedNamesInCtElementImpl_toString() throws Exception {
 		;
 		assertEquals( "the toString method of CtElementImpl should not shorten type names as it has no context or import statements", expected, computed );
 	}
+
+	@Test
+	public void printClassCreatedWithSpoon() throws Exception {
+
+		/* test that spoon is able to print a class that he created without setting manually the output (default configuration) */
+
+		Launcher l = new Launcher();
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/prettyprinter/DefaultPrettyPrinterTest.java]: Avoid **always** variables named like this!

File: src/main/java/spoon/support/DefaultCoreFactory.java [hunk 3]
```diff
@@ -277,6 +277,9 @@ public CtCatch createCatch() {
 		CtClass<T> e = new CtClassImpl<>();
 		e.setFactory(getMainFactory());
 		e.setParent(getMainFactory().Package().getRootPackage());
+		CompilationUnit cu = e.getFactory().CompilationUnit().create(e.getQualifiedName());
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/DefaultCoreFactory.java]: At this level, `e.getQualifiedName()` will return an empty string.

File: src/main/java/spoon/support/DefaultCoreFactory.java [hunk 4]
```diff
@@ -277,6 +277,9 @@ public CtCatch createCatch() {
...
+		cu.setDeclaredPackage(e.getPackage());
+		e.setPosition(new SourcePositionImpl(cu, -1, -1, -1, new int[0]));
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/DefaultCoreFactory.java]: Nice position... There isn't a better solution to fix the issue of @seintur?

File: src/test/java/spoon/test/prettyprinter/DefaultPrettyPrinterTest.java [hunk 5]
```diff
@@ -234,4 +239,33 @@ public void useFullyQualifiedNamesInCtElementImpl_toString() throws Exception {
...
+		CtClass<Object> ctClass = factory.Class().create("foo.Bar");
+		PrettyPrinter pp = new DefaultJavaPrettyPrinter(factory.getEnvironment());
+		JavaOutputProcessor jop =
+				new JavaOutputProcessor(File.createTempFile("foo","").getParentFile(),pp);
+		jop.setFactory(factory);
+		jop.createJavaFile(ctClass);
+		String pathname = System.getProperty("java.io.tmpdir") + "/foo/Bar.java";
+		assertTrue(new File(pathname).exists());
+
+		BufferedReader reader = new BufferedReader(new FileReader(pathname));
```
Comment (monperrus) [inline src/test/java/spoon/test/prettyprinter/DefaultPrettyPrinterTest.java]: use IOUtils?

**Pull request merged**